linera-service/src/grpc_proxy.rs

```diff
@@ -235,28 +236,31 @@ where
         health_reporter
             .set_serving::<ValidatorNodeServer<GrpcProxy<S>>>()
             .await;
-        let internal_server = Server::builder()
-            .add_service(self.as_notifier_service())
-            .serve(self.internal_address());
+        let (internal_server, _) = tasks.spawn_task(
+            Server::builder()
+                .add_service(self.as_notifier_service())
+                .serve(self.internal_address()),
+        );
         let reflection_service = tonic_reflection::server::Builder::configure()
             .register_encoded_file_descriptor_set(linera_rpc::FILE_DESCRIPTOR_SET)
             .build()?;
-        let public_server = self
-            .public_server()?
-            .layer(
-                ServiceBuilder::new()
-                    .layer(PrometheusMetricsMiddlewareLayer)
-                    .into_inner(),
-            )
-            .accept_http1(true)
-            .add_service(health_service)
-            .add_service(tonic_web::enable(self.as_validator_node()))
-            .add_service(tonic_web::enable(reflection_service))
-            .serve_with_shutdown(self.public_address(), shutdown_signal.cancelled_owned());
+        let (public_server, _) = tasks.spawn_task(
+            self.public_server()?
+                .layer(
+                    ServiceBuilder::new()
+                        .layer(PrometheusMetricsMiddlewareLayer)
+                        .into_inner(),
+                )
+                .accept_http1(true)
+                .add_service(health_service)
+                .add_service(tonic_web::enable(self.as_validator_node()))
+                .add_service(tonic_web::enable(reflection_service))
+                .serve_with_shutdown(self.public_address(), shutdown_signal.cancelled_owned()),
+        );
 
         select! {
-            internal_res = internal_server => internal_res?,
-            public_res = public_server => public_res?,
+            internal_res = internal_server => internal_res??,
+            public_res = public_server => public_res??,
         }
         Ok(())
```

Comment:
The point of using the `JoinSet` abstraction here is that we yield notional control over the execution of the spawned tasks to the user.  So here we should return a `Future` that joins the `JoinSet` so the user can decide when to block on it.

Reply:
I'm not sure I agree. The tasks are spawned so that they run on the background, possibly in different worker threads in the multi-threaded Tokio runtime.

The `JoinSet` here allows conditionally using the `LocalSet` when needed (TBH I'm not sure the gRPC proxy would ever need it, but at least it should be consistent with the other crates), while also cancelling all tasks when execution of one of the servers finish (or fail).